### ISSUE TITLE
unparsing is slow after #110 fix

### ISSUE THREAD

**Issue by denis-bogdanas** [runtimeverification/k#121]
Pretty printing cells k and the constraint inside a proof took <0.1s before, ~40s after.

Consider supporting 2 versions of unparsing: the old and the new one.

This script illustrates the issue. It runs a relatively short proof from casper project, on 2 versions of K `gnosis` branch. One before the rebase on latest master, one after. Run the script below inside an empty dir, go to `verified-smart-contracts` and compare the two `.out` files. Compare the latest 2 logged times: `STEP 2724 v1` and `TOTAL TIME:`.

```
#!/usr/bin/env bash
git clone --recurse-submodules https://github.com/runtimeverification/verified-smart-contracts.git
cd verified-smart-contracts/.build/evm-semantics/.build/k
git fetch
git checkout gnosis_slow_pp
mvn package -Dmaven.test.skip=true
cd ../..
make defn
.build/k/k-distribution/target/release/k/bin/kompile --debug --backend java -I .build/java -d .build/java --main-module ETHEREUM-SIMULATION --syntax-module ETHEREUM-SIMULATION .build/java/driver.k
cd ../..
make casper
.build/evm-semantics/.build/k/k-distribution/target/release/k/bin/kprove specs/casper/total_curdyn_deposits_scaled-success-spec.k -d .build/evm-semantics/.build/java -m VERIFICATION \
        --z3-executable --log-basic> pretty-print-k-and-contraint-slow.out 2>&1

cd .build/evm-semantics/.build/k
git checkout gnosis_before_rebase
mvn package -Dmaven.test.skip=true
cd ../..
.build/k/k-distribution/target/release/k/bin/kompile --debug --backend java -I .build/java -d .build/java --main-module ETHEREUM-SIMULATION --syntax-module ETHEREUM-SIMULATION .build/java/driver.k
cd ../..
.build/evm-semantics/.build/k/k-distribution/target/release/k/bin/kprove specs/casper/total_curdyn_deposits_scaled-success-spec.k -d .build/evm-semantics/.build/java -m VERIFICATION \
        --z3-executable --log-basic> pretty-print-k-and-contraint-fast.out 2>&1

```

**Comment by denis-bogdanas** [comment]
@dwightguth Can you look at this please? Vyper compiler runs twice slower now. I cannot run the whole test suite even if I leave it overnight.

**Comment by dwightguth** [comment]
I really need a better example if I'm going to test this; you're asking me to check out some random branch of k that may or may not contain changes that will conflict with my ability to debug this issue. Give me a minimal example that I can test with the master branch of k and I'll look into it.

**Comment by denis-bogdanas** [comment]
The changes are definitely related to #110. I already verified this. You can pick any semantics from tutorial and run it with master and pre-#110 k. Preferably one with a bigger configuration. Even for IMP I noticed that pretty printing adds a few seconds.

**Comment by dwightguth** [comment]
I tested kool-typed-dynamic with the latest version of k and was unable to reproduce any kind of performance regression. I did however reproduce it with commit a3adf2120bde353b7cd89f2d31662ff2c24c6631 that you mentioned, so I suspect this relates to the performance problems associated with Module.overloads() which was fixed by #125.

Please close if updating to the latest version of k fixes your issue, otherwise I will need a more specific example that you have tested on the latest version of k.

**Comment by denis-bogdanas** [comment]
Finally got time to verify this. Pretty printing is faster now, around ~0.5s overhead for the example I posted but more for real scenarios we are using with kprove. It is still slower than before the changes, up to 10x slower. I cannot measure exactly because many things since changed.

We'll live with that. Thanks for looking into this!
